File: haskell/lista11/zad6.hs
module Test where

data Tree a = Node (Tree a) a (Tree a) | Leaf

insert Leaf a = Node Leaf a Leaf
insert (Node left key right) a =
	if a < key then Node (insert left a) key right
	else Node left key (insert right a)

flatten t =
	let
		flattenH Leaf acc = acc
		flattenH (Node left k right) acc =
			flattenH right (k : flattenH left acc)
	in reverse (flattenH t [])
	
treeSort ls =
	flatten (foldl insert Leaf ls)

File: prolog/lista5/zad6.pl
split([], _, [], []).
split([H|T], E, [H|S], B) :- H < E, split(T, E, S, B).
split([H|T], E, S, [H|B]) :- H >= E, split(T, E, S, B).

qsort(X, Y) :- qsort(X, [], R), reverse(R, Y).

qsort([], X, X).
qsort([H], X, [H|X]) :- !.
qsort([H|T], ACC, R) :- split(T, H, S, B), qsort(S, ACC, R1), qsort(B, [H|R1], R).
